source/common/stats/heap_stat_data.cc

```diff
@@ -60,61 +49,85 @@ void HeapStatDataAllocator::debugPrint() {
+  StatsSharedImpl(StatName name, HeapStatDataAllocator& alloc, absl::string_view tag_extracted_name,
+                  const std::vector<Tag>& tags)
+      : MetricImpl<BaseClass>(name, tag_extracted_name, tags, alloc.symbolTable()), alloc_(alloc) {}
 
-  ~CounterImpl() override {
+  ~StatsSharedImpl() override {
     // MetricImpl must be explicitly cleared() before destruction, otherwise it
     // will not be able to access the SymbolTable& to free the symbols. An RAII
     // alternative would be to store the SymbolTable reference in the
     // MetricImpl, costing 8 bytes per stat.
-    MetricImpl::clear();
-    alloc_.removeCounterFromSet(this);
-    data_.free(symbolTable());
+    this->clear(symbolTable());
   }
 
-  // Stats::Counter
-  void add(uint64_t amount) override {
-    data_.value_ += amount;
-    data_.pending_increment_ += amount;
-    data_.flags_ |= Flags::Used;
-  }
-  void inc() override { add(1); }
-  uint64_t latch() override { return data_.pending_increment_.exchange(0); }
-  void reset() override { data_.value_ = 0; }
-  bool used() const override { return data_.flags_ & Flags::Used; }
-  uint64_t value() const override { return data_.value_; }
-
+  // Metric
   SymbolTable& symbolTable() override { return alloc_.symbolTable(); }
-  StatName statName() const override { return data_.statName(); }
+
+  // Counter/Gauge
+  bool used() const override { return flags_ & Metric::Flags::Used; }
 
   // RefcountInterface
-  void incRefCount() override { ++data_.ref_count_; }
+  void incRefCount() override { ++ref_count_; }
   bool decRefCount() override {
-    ASSERT(data_.ref_count_ >= 1);
-    return --data_.ref_count_ == 0;
+    ASSERT(ref_count_ >= 1);
+    return --ref_count_ == 0;
   }
-  uint32_t use_count() const override { return data_.ref_count_; }
+  uint32_t use_count() const override { return ref_count_; }
 
-private:
-  HeapStatData& data_;
+protected:
   HeapStatDataAllocator& alloc_;
+
+  // Holds backing store for both CounterImpl and GaugeImpl. This provides a level
+  // of indirection needed to enable stats created with the same name from
+  // different scopes to share the same value.
+  std::atomic<uint64_t> value_{0};
+  std::atomic<uint64_t> pending_increment_{0};
+  std::atomic<uint16_t> flags_{0};
+  std::atomic<uint16_t> ref_count_{0};
 };
 
-class GaugeImpl : public Gauge, public MetricImpl {
+class CounterImpl : public StatsSharedImpl<Counter> {
 public:
-  GaugeImpl(HeapStatData& data, HeapStatDataAllocator& alloc, absl::string_view tag_extracted_name,
+  CounterImpl(StatName name, HeapStatDataAllocator& alloc, absl::string_view tag_extracted_name,
+              const std::vector<Tag>& tags)
+      : StatsSharedImpl(name, alloc, tag_extracted_name, tags) {}
+  ~CounterImpl() override { alloc_.removeCounterFromSet(this); }
+
+  // Stats::Counter
+  void add(uint64_t amount) override {
+    value_ += amount;
```

Comment:
These are good points. My view on this as always been that this is eventually consistent and should favor performance. It's possible that things have gotten more complicated since used was originally created a long time ago, so I agree it's worth it to document this here or elsewhere. Maybe add a bunch of comments that we can revisit later if needed?